modules/node-shared/src/main/scala/org/tessellation/node/shared/infrastructure/snapshot/GlobalSnapshotStateChannelEventsProcessor.scala

```diff
@@ -67,26 +66,64 @@ object GlobalSnapshotStateChannelEventsProcessor {
     new GlobalSnapshotStateChannelEventsProcessor[F] {
       private val logger = Slf4jLogger.getLoggerFromClass[F](GlobalSnapshotStateChannelEventsProcessor.getClass)
 
+      def deserialize[A: Decoder](binary: Signed[StateChannelSnapshotBinary]): F[Option[A]] =
+        jsonBrotliBinarySerializer.deserialize[A](binary.value.content).map(_.toOption)
+
+      def buildSnapshotFeesInfo(
+        lastGlobalSnapshotInfo: GlobalSnapshotInfo,
+        event: StateChannelOutput,
+        allFeesAddresses: Map[Address, Set[Address]]
+      ): F[SnapshotFeesInfo] =
+        event.snapshotBinary.value.lastSnapshotHash match {
+          case hash if hash == Hash.empty => SnapshotFeesInfo.empty.pure // genesis
+          case _ =>
+            deserialize[Signed[CurrencyIncrementalSnapshot]](event.snapshotBinary).flatMap {
+              case None => new Exception("Could not get snapshot after deserializing").raiseError[F, SnapshotFeesInfo]
+              case Some(snapshot) =>
+                Async[F].delay {
+                  val stakingBalance = fetchStakingBalance(event.address, lastGlobalSnapshotInfo)
+                  val sortedMessagesDesc = snapshot.value.messages.map(_.toList.sortBy(_.ordinal.value.value)(Ordering[Long].reverse))
```

Comment:
As a follow up to https://github.com/Constellation-Labs/tessellation/pull/890#discussion_r1670923990

```scala
@derive(eqv, show, encoder, decoder, order, ordering)
case class CurrencyMessage(messageType: MessageType, address: Address, metagraphId: Address, parentOrdinal: MessageOrdinal) {
    def ordinal: MessageOrdinal = parentOrdinal.next
}
```

In such code the `Ordering[CurrencyMessage]` sorts by all the fields... instead of just by `ordinal`. That being said, the `SortedSet[Signed[CurrencyMessage]]` is wrongly sorted in snapshots because the wrong `Ordering` was provided in `CurrencyMessage`. Of course `SortedSet` makes sure that all the nodes have it sorted in the same way but in all the places where you want to either get first or last ordinal you can't simply rely on `head` or `last`. It is a little bit error-prone.

The worst part is that even if you provide the correct one now globally:

```scala
  @derive(eqv, show, encoder, decoder)
  case class CurrencyMessage(messageType: MessageType, address: Address, metagraphId: Address, parentOrdinal: MessageOrdinal) {
    def ordinal: MessageOrdinal = parentOrdinal.next
    implicit val order: Order[CurrencyMessage] = Order[MessageOrdinal].contramap(_.ordinal)
    implicit val ordering: Ordering[CurrencyMessage] = Order[CurrencyMessage].toOrdering
  }
```

Then new snapshots will be fine but old ones will be incorrect because these were sorted in an old way so hash is calculated in a different way...

So I guess we can't fix it without wiping the testnet/integrationnnet chain and we need to sort it locally.

So TL:DR I am fine with sorting it there unless we verify if we can fix it globally.

cc @sdavidp

Reply:
Got it, so let's keep locally for now